crates/core/src/nats/client.rs

```diff
@@ -0,0 +1,223 @@
+use std::{sync::Arc, time::Duration};
+
+use bytes::Bytes;
+use tracing::info;
+
+use super::{
+    types::{
+        JetStreamConfig,
+        JetStreamContext,
+        NatsConsumer,
+        NatsStream,
+        PullConsumerConfig,
+    },
+    NatsError,
+    Subject,
+};
+use crate::types::BoxedResult;
+
+#[derive(Debug, Clone)]
+pub struct NatsClient {
+    pub url: String,
+    pub conn_id: String,
+    pub conn: Arc<async_nats::Client>,
+    pub jetstream: Arc<JetStreamContext>,
+}
+
+impl NatsClient {
+    pub async fn connect(
+        url: &str,
+        conn_id: &str,
+        nkey: Option<String>,
+    ) -> Result<Self, NatsError> {
+        let conn_id = conn_id.to_string();
+        let conn = Self::create_conn(url, nkey).await?;
+        let context =
+            Arc::new(async_nats::jetstream::new(conn.as_ref().clone()));
+
+        info!("Connected to NATS server at {}", url);
+        Ok(Self {
+            url: url.to_string(),
+            conn_id,
+            conn,
+            jetstream: context,
+        })
+    }
+
+    async fn create_conn(
+        url: &str,
+        nkey: Option<String>,
+    ) -> Result<Arc<async_nats::Client>, NatsError> {
+        let options = async_nats::ConnectOptions::new()
+            .connection_timeout(Duration::from_secs(30))
+            .max_reconnects(10);
+
+        if nkey.is_none() {
+            return Err(NatsError::NkeyNotProvided {
+                url: url.to_owned(),
+            })
+        }
+
+        let conn =
+            async_nats::connect_with_options(&url, options.nkey(nkey.unwrap()))
+                .await
+                .map_err(|e| NatsError::ConnectError {
+                    url: url.to_owned(),
+                    source: e,
+                })?;
+
+        Ok(Arc::new(conn))
+    }
+
+    pub async fn validate_payload(
+        &self,
+        payload: Bytes,
+        subject_name: &str,
+    ) -> Result<&Self, NatsError> {
+        let payload_size = payload.len();
+        let max_payload_size = self.conn.server_info().max_payload;
+        if payload_size > max_payload_size {
+            return Err(NatsError::PayloadTooLarge {
+                subject: subject_name.to_string(),
+                payload_size,
+                max_payload_size,
+            });
+        }
+
+        Ok(self)
+    }
+
+    pub async fn publish(
+        &self,
+        subject: Subject,
+        payload: Bytes,
+    ) -> BoxedResult<&Self> {
+        let subject_prefixed = subject.with_prefix(&self.conn_id);
+        let context = self.jetstream.as_ref();
+        let ack_future = context.publish(subject_prefixed, payload).await?;
```

Comment:
Sorry, what I meant was that we can get rid of `validate_payload`, and then match on this future's error to return log an error if the payload is too large.

Reply:
There's no payload size error related to the publish futures, I don't see this as a huge resource-consuming or complex algorithm, it's not introducing high level of complexity or something like that, this could be improved to have `max_payload_size` in the struct to avoid get this from the server each time, but anyway, isn't something to go further, I don't see problems to stay with it here

![CleanShot 2024-07-24 at 17 38 10](https://github.com/user-attachments/assets/072fb71a-698a-41aa-9557-76d22cb4cd02)

Reply:
**tldr;** 
I tested here, and it's impossible to do what you propose. In fact, it's better to always check the payload using the `publish` method since it's just a simple conditional check.

First, as you can see, there's no error related to payload size in the publish future's error enum:
![CleanShot 2024-07-24 at 17 56 13](https://github.com/user-attachments/assets/324a9b46-765b-435a-9d1f-9cd17262026a)

Second, if you try to decrease the `max_payload` in the `nats.conf`, you will see that by doing that and publishing anything, there's no "max payload error"; the error that shows on Nats is a timeout error as below.

```
**Error: CreateStreamFailed { name: "connection-1437309941_stream:blocks", source: Error { kind: TimedOut, source: None } }**
```

Apart from these points, as mentioned earlier, I don't see a major issue here because this method is quite simple. There's no asynchronous processing involved (please refer to the most updated version), and the amount of processing is minimal, with almost no added complexity. I don't foresee this becoming a performance issue in the future. However, we did encounter difficulties with this validation when Maciej was testing, which is why he introduced this method. If we have concerns, we can review this and explore potential solutions. For now, in my humble opinion, we can proceed as planned.

Reply:
> Second, if you try to decrease the max_payload in the nats.conf, you will see that by doing that and publishing anything, there's no "max payload error"; the error that shows on Nats is a timeout error as below.

Yes, that still works. We can match on the `TimedOut` error and then check if the payload was greater than our max payload to deduce what gets logged. 

>  I don't foresee this becoming a performance issue in the future. However, we did encounter difficulties with this validation when Maciej was testing, which is why he introduced this method. If we have concerns, we c

This is more of a readability concern than a performance optimization. The success path for publishing should be very minimal when read making it easy to point out exceptions.

> we can review this and explore potential solutions. For now, in my humble opinion, we can proceed as planned.

Yes, I have no worries merging this. Just wanted to emphasize the preferred paradigm of avoiding defensive programming until when necessary (i.e. for unstructured inputs).